src/modules/jcms_notifications/src/ArticlePresave.php

```diff
@@ -0,0 +1,164 @@
+ */
+class ArticlePresave {
+
+  /**
+   * @var \Drupal\jcms_notifications\FetchArticleService
+   */
+  protected $fetchArticle;
+
+  /**
+   * @var \Drupal\Core\Entity\EntityInterface
+   */
+  protected $node;
+
+  /**
+   * @var null
+   */
+  protected static $articleData = NULL;
+
+  /**
+   * ArticlePresave constructor.
+   *
+   * @param \Drupal\jcms_notifications\FetchArticleService $fetch_article
+   */
+  public function __construct(FetchArticleService $fetch_article) {
+    $this->fetchArticle = $fetch_article;
+  }
+
+  /**
+   * Sets the node property.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   */
+  public function setNode(EntityInterface $entity) {
+    $this->node = $entity;
+  }
+
+  /**
+   * Gets the node property.
+   *
+   * @return \Drupal\Core\Entity\EntityInterface
+   */
+  public function getNode() {
+    return $this->node;
+  }
+
+  /**
+   * Returns true if the node is an article.
+   *
+   * @return bool
+   */
+  public function isArticle() {
+    return ($this->node->bundle() == 'article');
+  }
+
+  /**
+   * Returns true if the node is new.
+   *
+   * @return bool
+   */
+  public function isNew() {
+    return $this->node->isNew();
+  }
+
+  /**
+   * Checks if the node is valid.
+   *
+   * @return bool
+   */
+  public function nodeIsValid() {
+    if (!($this->getNode() instanceof EntityInterface)) {
+      return FALSE;
+    }
+    if (!$this->isArticle()) {
+      return FALSE;
+    }
+    if ($this->isNew()) {
+      return FALSE;
+    }
+    return TRUE;
+  }
+
+  /**
+   * Gets the article ID from the node.
+   *
+   * @return null|string
+   */
+  public function getArticleIdFromNode() {
+    return $this->getNode()->label();
+  }
+
+  /**
+   * Gets the article data.
+   *
+   * @param $id
+   *
+   * @return array|null
+   */
+  public function getArticleById($id) {
+    if (self::$articleData !== NULL) {
+      self::$articleData = $this->fetchArticle->getArticle($id);
+    }
+    return self::$articleData;
+  }
+
+  /**
+   * Adds the JSON fields to the node.
+   */
+  public function addJsonFields() {
+    if ($this->nodeIsValid()) {
+      $id = $this->getArticleIdFromNode();
+      $article = $this->getArticleById($id);
+      $pid = $this->getNode()->get('field_article_json')->getValue()[0]['target_id'];
+      $paragraph = Paragraph::load($pid);
+      $paragraph->set('field_article_unpublished_json', $article['data']['unpublished']);
+      if ($article['data']['published']) {
+        $paragraph->set('field_article_published_json', $article['data']['published']);
+      }
+    }
+  }
+
+  /**
+   * Sets the published date on the node.
+   */
+  public function setPublishedDate() {
+    if ($this->nodeIsValid()) {
+      $id = $this->getArticleIdFromNode();
+      $article = $this->getArticleById($id);
+      // Whether new or not, set the published date.
+      $json = json_decode($article['data']['unpublished']);
+      $date = date('U', $json['published']);
```

Comment:
We've had some trouble with dates previously, might be worth checking if `$date` comes out as a `bool`